internal/packer_registry/types.bucket.go

```diff
@@ -0,0 +1,387 @@
+func (b *Bucket) Initialize(ctx context.Context) error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
+
+	if b.client == nil {
+		if err := b.connect(); err != nil {
+			return err
+		}
+	}
+
+	bucketInput := &models.HashicorpCloudPackerCreateBucketRequest{
+		BucketSlug:  b.Slug,
+		Description: b.Description,
+		Labels:      b.Labels,
+	}
+
+	err := UpsertBucket(ctx, b.client, bucketInput)
+	if err != nil {
+		return fmt.Errorf("failed to initialize iteration for bucket %q: %w", b.Slug, err)
+	}
+
+	// Create/find iteration
+	iterationInput := &models.HashicorpCloudPackerCreateIterationRequest{
+		BucketSlug:  b.Slug,
+		Fingerprint: b.Iteration.Fingerprint,
+	}
+
+	var existingBuilds []*models.HashicorpCloudPackerBuild
+
+	id, err := CreateIteration(ctx, b.client, iterationInput)
+	if err != nil {
+		if !checkErrorCode(err, codes.AlreadyExists) {
+			return fmt.Errorf("failed to create Iteration for Bucket %s with error: %w", b.Slug, err)
+		} else {
+			// load existing iteration using fingerprint.
+			id, err = GetIteration(ctx, b.client, b.Slug, b.Iteration.Fingerprint)
+			if err != nil {
+				return fmt.Errorf("Error loading the existing iteration for fingerprint %s", b.Iteration.Fingerprint)
+			}
+			log.Println("[TRACE] a valid iteration was retrieved with the id", id)
+			// list all this iteration's builds so we can figure out which ones
+			// we want to run against. TODO: pagination?
+			existingBuilds, err = ListBuilds(ctx, b.client, b.Slug, id)
+			if err != nil {
+				return fmt.Errorf("Error listing builds for this existing iteration: %s", err)
+			}
+		}
+	} else {
+		log.Println("[TRACE] a valid iteration for build was created with the Id", id)
+	}
+
+	b.Iteration.ID = id
+
+	var errs *multierror.Error
+	var wg sync.WaitGroup
+
+	// remove the iteration's expected builds that already exist.
+	toCreate := []string{}
+	for _, existing := range existingBuilds {
+		found := false
+		for i, expected := range b.Iteration.expectedBuilds {
+			if existing.ComponentType == expected {
+				log.Printf("build of component type %s already exists; skipping the create call", expected)
+				found = true
+				if *existing.Status == models.HashicorpCloudPackerBuildStatusDONE {
+					// We also need to remove the builds that are _complete_ from the
+					// Iteration's expectedBuilds so we don't overwrite them.
+					b.Iteration.expectedBuilds = append(b.Iteration.expectedBuilds[:i],
+						b.Iteration.expectedBuilds[i+1:]...)
+					log.Printf("build of component type %s is already marked DONE; "+
+						"removing the build from the HCP Packer Registry expected builds.", expected)
+				}
+				break
+			}
+		}
+		if !found {
+			toCreate = append(toCreate, existing.ComponentType)
+		}
+	}
+
+	for _, buildName := range toCreate {
+		wg.Add(1)
+		go func(name string) {
+			defer wg.Done()
+			log.Printf("[TRACE] registering build with iteration for %q.", name)
+			// Need a way to handle skipping builds that were already created.
```

Comment:
I thought the `toCreate` slice would only have non-created builds. Is this comment outdated?

Reply:
Good catch.. it is being resolved in the next commit that I am working on locally.